### PR TITLE
Pybind secondaries

### PR DESCRIPTION
**Pull request by MaxSac**

### CODE REVIEW

**Comment by sudojan** [comment]
It would be helpful if, before the review, no auto formatting is applied.
Otherwise most of the committed lines (already excluding white spaces) are just line breaks.
It is just tedious and unnecessary extrawork to search for the real changes

File: src/pyPROPOSAL/detail/parametrization.cxx [hunk 1]
```diff
@@ -144,57 +141,48 @@ void init_parametrization(py::module& m)
...
             Get internal hash corresponding to the current parametrization
     )pbdoc";
 
-    py::class_<crosssection::Parametrization<Medium>,
-        std::shared_ptr<crosssection::Parametrization<Medium>>>(
-        m_sub, "ParametrizationForMedium", param_docstring_class)
+template <typename T>
+void decl_param(pybind11::module m_sub, std::string const& class_name)
+{
+    pybind11::class_<crosssection::Parametrization<T>,
```
Comment (Jean1995) [inline src/pyPROPOSAL/detail/parametrization.cxx:151]: these `pybind11` could all be just `py` to have shorter code

File: src/pyPROPOSAL/detail/parametrization.cxx [hunk 2]
```diff
@@ -128,7 +125,7 @@ void init_parametrization(py::module& m)
             e.g. the differential crosssection in v for the current
             parametrization multipied by :math:`v^2`.  If the parametrization is
             given in a double-differential-crosssection the crosssection will be
-            integrated over all values but v.
+         1  1ntegrated over all values but v.
```
Comment (Jean1995) [inline src/pyPROPOSAL/detail/parametrization.cxx:128]:
```suggestion
            integrated over all values but v.
```

File: src/pyPROPOSAL/detail/secondaries.cxx [hunk 3]
```diff
@@ -16,52 +18,114 @@
 namespace py = pybind11;
 using namespace PROPOSAL;
 
+template <typename T, typename BaseT> struct SecondariesBuilder {
+    template <typename... Args> auto decl_virtuell_param(Args... args)
```
Comment (Jean1995) [inline src/pyPROPOSAL/detail/secondaries.cxx:22]: virtuell -> virtual

File: src/pyPROPOSAL/detail/secondaries.cxx [hunk 4]
```diff
@@ -16,52 +18,114 @@
...
+    SecondariesBuilder<secondaries::Compton, secondaries::Parametrization> {}
+        .decl_virtuell_param(m_sub, "Compton");
+    SecondariesBuilder<secondaries::EpairProduction,
+        secondaries::Parametrization> {}
+        .decl_virtuell_param(m_sub, "EpairProduction");
+    SecondariesBuilder<secondaries::Ionization, secondaries::Parametrization> {}
+        .decl_virtuell_param(m_sub, "Ionization");
+    SecondariesBuilder<secondaries::MupairProduction,
+        secondaries::Parametrization> {}
+        .decl_virtuell_param(m_sub, "MupairProduction");
```
Comment (Jean1995) [inline src/pyPROPOSAL/detail/secondaries.cxx:70]: Shouldn't this also be `decl_rho_param`?

File: src/pyPROPOSAL/detail/parametrization.cxx (hunk 1 above)
Comment (MaxSac) [reply src/pyPROPOSAL/detail/parametrization.cxx:151]: :+1:

File: src/pyPROPOSAL/detail/secondaries.cxx (hunk 4 above)
Comment (MaxSac) [reply src/pyPROPOSAL/detail/secondaries.cxx:70]: yes

Comment (Jean1995) [reply src/pyPROPOSAL/detail/secondaries.cxx:70]: Oh, no, it does not, my mistake :D

File: src/pyPROPOSAL/detail/secondaries.cxx [hunk 5]
```diff
@@ -16,52 +18,114 @@
...
+            py::init<ParticleDef, Medium>(), py::arg("particle_def"),
+            py::arg("medium"));
+    }
+
+    template <typename... Args> auto decl_rho_param(Args... args)
+    {
+        py::class_<T, BaseT, std::shared_ptr<T>>(args...)
+            .def(py::init<ParticleDef, Medium>(), py::arg("particle_def"),
+                py::arg("medium"))
+            .def("calculate_rho", &T::CalculateRho);
```
Comment (sudojan) [inline src/pyPROPOSAL/detail/secondaries.cxx:39]: Maybe a docstring explaining what rho is would be helpful

**Review by sudojan — APPROVED** [review]

**Pull request merged**